File: 0018_4Sum/Solution2.py
from typing import *

class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        
        def n_sum(N, nums, target, path, res):
            if (N < 2 or
                len(nums) < N or
                nums[0] * N > target or
                nums[-1] * N < target):
                return
            
            if N > 2:
                for i in range(len(nums) + 1 - N):
                    if i > 0 and nums[i] == nums[i-1]: continue
                    n_sum(N-1, nums[i+1:], target-nums[i], path + [nums[i]], res)
                return

            l, r = 0, len(nums) - 1
            while l < r:
                _sum = nums[l] + nums[r]
                if _sum < target:
                    l += 1
                elif _sum > target:
                    r -= 1
                else:
                    res.append(path + [nums[l], nums[r]])
                    l += 1
                    r -= 1
                    while l < r and nums[l-1] == nums[l]: l += 1
                    while l < r and nums[r+1] == nums[r]: r -= 1

        nums.sort()
        res = []
        n_sum(4, nums, target, [], res)
        return res


def test(test_name, nums, target, expected):
    res = Solution().fourSum(nums, target)
    [item.sort() for item in res]
    [item.sort() for item in expected]
    res.sort()
    expected.sort()

    if res == expected:
        print(test_name + ' success.')
    else:
        print(test_name + ' failed.')


if __name__ == '__main__':
    nums1 = [1, 0, -1, 0, -2, 2]
    target1 = 0
    expected1 = [
        [-1,  0, 0, 1],
        [-2, -1, 1, 2],
        [-2,  0, 0, 2]
    ]
    test('test1', nums1, target1, expected1)

    nums2 = [-2,-1,-1,1,1,2,2]
    target2 = 0
    expected2 = [
        [-2,-1,1,2],
        [-1,-1,1,1]
    ]
    test('test2', nums2, target2, expected2)

File: of03_数组中重复的数字/Solution2.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int findRepeatNumber(vector<int>& nums) {
        unordered_map<int, int> mp;
        for (int i = 0; i < nums.size(); i++) {
            if (mp.find(nums[i]) != mp.end())
                return nums[i];
            mp[nums[i]]++;
        }
        return -1;
    }
};

void test(string test_name, vector<int> nums, vector<int> expected)
{
    int res = Solution().findRepeatNumber(nums);
    if (find(expected.begin(), expected.end(), res) != expected.end())
        cout << test_name << " success." << endl;
    else
        cout << test_name << " failed." << endl;
}

int main()
{
    vector<int> nums1 = {2, 3, 1, 0, 2, 5, 3};
    vector<int> expected1 = {2,3};
    test("test1", nums1, expected1);

    return 0;
}

// [2, 3, 1, 0, 2, 5, 3]
// 输出：2 或 3 

// 在一个长度为 n 的数组 nums 里的所有数字都在 0～n-1 的范围内。
// 数组中某些数字是重复的，但不知道有几个数字重复了，也不知道每个数字重复了几次。
// 请找出数组中任意一个重复的数字。

// 限制：

// 2 <= n <= 100000
